`tval` command emulates node with es6 inputs.

johnjbarton opened the pull request:
./tval program args
runs program JS file with argv [path-to-tval, 'program', ...args]

arv commented on node6:
```diff
@@ -0,0 +1,37 @@
+// normalized module name.
+var moduleName = filename.replace(/\\/g, '/');
+
+// Simuulate a future where it looks like node is executing our program
```
typo

johnjbarton replied:
done

arv commented:
I like. We used to do this for `traceur` before it turned into a kitchen sink.

Is node6 the right name for this? Maybe traceur-node?

@domenic For node expert input.

domenic commented on node6:
```diff
@@ -0,0 +1,37 @@
+var moduleName = filename.replace(/\\/g, '/');
+
+// Simuulate a future where it looks like node is executing our program
+process.argv = ['node'].concat(process.argv.slice(2));
```
I would do `node6` instead of `node` here. Or maybe read from `process.execPath`, since people might have invoked it with any variety of (node|iojs)[.exe]

johnjbarton replied:
done

domenic commented:
Does this support using `require` seamlessly from inside ES6 code? What about `import`? What about `require`ing Traceur-authored code or `import`ing CommonJS code?

I think at a minimum it needs to support `require`ing CommonJS code and `import`ing ES6 code.

My approach with this was [traceur-runner](https://github.com/domenic/traceur-runner) which doesn't do the arguments stuff but does focus on module system interop.

johnjbarton commented:
> Is node6 the right name for this? Maybe traceur-node?
tnode? I'd prefer shorter and w/o punctuation. 

> Does this support using require seamlessly from inside ES6 code?

Here is my understanding: 

If `options.require` is true and `modules=register`, yes. In that case the modules are instantiated with an argument `require` that resets the path to correct for bundling. 
If `modules=commonjs` then yes of the modules are left in the same directory relationship as the inputs.
If the `require` argument is a package address, yes.

> What about import? 
Of course, that's the point.

> What about requireing Traceur-authored code 
This will work if the code was compiled with `commonjs` and the tree was not changed.

> or importing CommonJS code?
This will work if the developer figures out what `default` does and uses it correctly and we are lucky.

This commit does not change anything about how traceur works with `import` or `require()`. It only changes the command line available for interpretation with traceur.

domenic commented:
My point was that I don't think this is going to be a very useful command for anything but Traceur's own makefile unless the interop story is good. Based on your responses above it sounds like maybe tnode should set some options up ahead of time to accomplish this. I think some tests to demonstrate the interop story would be important.

Otherwise, if this is purely about patching argv, don't give it a name that implies that it integrates well with node. Just call it `argv-patcher` or something.

But maybe I am misinterpreting the intent of this PR and indeed the point is just to create something to help the makefile, and not to create a command that you anticipate other developers actually using.

johnjbarton commented:
My previous assessment was off base. This command will "simulate node under es6" for the most common cases:
  * importing es6 code from es6 code,
  * require() es5 npm packages from es6 code.

I believe that is how most projects will work. Some projects may also need
  * require() es5 not in packages from es6 code.

I think this is an unwise strategy and other tools exist that support that scenario in any case. More projects may want
  * create es5 npm packages from es6 code,

but that is transpilation not execution. 

While I believe this command is generally useful I'm not personally planning on investigating the interop issues. I've renamed the command `tval` to avoid confusion with `*node`. 

I added a usage() as well. PTAL

arv commented on test/unit/node/generated-code-dependencies.js:
```diff
@@ -524,4 +524,12 @@ suite('context test', function() {
+    var cmd = './tval test/unit/node/resources/test_tval.js --arg1 --arg2 arg3';
+    exec(cmd, function(error, stdout, stderr) {
+      assert.isNull(error);
+      assert.equal(stdout, 'args--arg1--arg2arg3\n');
```
Does this need a `done` since it is async?

johnjbarton replied:
done

arv commented on tval:
```diff
@@ -0,0 +1,46 @@
+var filename = process.argv[2];
+
+if (!filename || filename === '-?' || filename === '--help') {
+	usage();
```
indentation error

johnjbarton replied:
done

arv commented on tval:
```diff
@@ -0,0 +1,46 @@
+// The first argument is 'node6'; the second one is our 'program' name.
+var filename = process.argv[2];
+
+if (!filename || filename === '-?' || filename === '--help') {
```
`-h` too?

johnjbarton replied:
done

arv commented:
I think this command is very useful, however I don't understand the name `tval`? Maybe something with interpret, which the old traceur command can also do but it does it wrong.

johnjbarton commented:
Meant to rhyme with 'eval' ;-) 
`trapert`? `intraceur` 'ti` `intr` `trintr`?
`kong` http://www.urbandictionary.com/define.php?term=Kong
Any barn color that is short and w/o punctuation.

johnjbarton commented:
PTAL, left the name pending feedback.

arv commented:
LGTM, tval is fine.

arv commented:
Don't forget to update the commit description.
